### ISSUE TITLE
Property access from edge/vertex in array gives unexpected error

### ISSUE THREAD

**Issue by MuhammadTahaNaveed** [apache/age#1033]
**Describe the bug**
Property access from edge/vertex in array gives unexpected error.

**How are you accessing AGE (Command line, driver, etc.)?**
- terminal

**What data setup do we need to do?**
```pgsql
SELECT * FROM cypher('graph',$$
    CREATE (n)-[:KNOWS {n:'hello'}]->({n:'hello'})
$$) as (a agtype);
```

**What is the command that caused the error?**
```pgsql
SELECT * FROM cypher('graph',$$
    MATCH ()<-[x *]-()
    RETURN x[0].n
$$) as (a agtype);
```
```
ERROR:  container must be an array or object
```

**Expected behavior**
should have executed successfully and returned `hello`

**Environment (please complete the following information):**
- 1.3.0 PG12

**Comment by Amr-Shams** [comment]
can you explain more, cause the only case here is when I do have edges and their proeprites are not defined for the sense of accessing non exist value the return error should be valid. 
can you pleas share with us the x as an array. that might help to manually test

**Comment by MuhammadTahaNaveed** [comment]
VLE gives an array of edges. For example
```
SELECT * FROM cypher('graph',$$MATCH ()<-[x *]-() RETURN x$$) as (a agtype);
                                                                    a                                                                     
------------------------------------------------------------------------------------------------------------------------------------------
 [{"id": 844424930131969, "label": "KNOWS", "end_id": 281474976710658, "start_id": 281474976710657, "properties": {"n0": "hello"}}::edge]
```
Now if we try to access a certain property from first edge in array, I should be able to do the query expecting the following result
```
age=# SELECT * FROM cypher('graph',$$MATCH ()<-[x *]-() RETURN x[0].n0$$) as (a agtype);
    a    
---------
 "hello"
(1 row)

```

But instead, it errors out.
```
age=# SELECT * FROM cypher('graph',$$MATCH ()<-[x *]-() RETURN x[0].n0$$) as (a agtype);
ERROR:  container must be an array or object
```
I hope this explains.

**Comment by jrgemignani** [comment]
I think the misunderstanding is that the VLE component doesn't return an array of edges, it returns a row for each edge found.

```
psql-13.9-5432-pgsql=# SELECT * FROM cypher('graph',$$ MATCH ()<-[x *]-() RETURN x $$) as (a agtype);
                                                                    a

------------------------------------------------------------------------------------------------------------------------------------
-----
 [{"id": 844424930131969, "label": "KNOWS", "end_id": 281474976710658, "start_id": 281474976710657, "properties": {"n": "hello"}}::e
dge]
 [{"id": 844424930131970, "label": "KNOWS", "end_id": 281474976710660, "start_id": 281474976710659, "properties": {"n": "hello"}}::e
dge]
(2 rows)

psql-13.9-5432-pgsql=#
```

Each edge is an object, not an array. Although, it is wrapped in an array, it is still an object. So, only object access operators will work on it.

**Comment by jrgemignani** [comment]
There does seem to be an issue with how `x` is resolved for vle and for non-vle -

```
psql-13.9-5432-pgsql=# SELECT * FROM cypher('graph',$$ MATCH ()<-[x *]-() RETURN x.name $$) as (a agtype);
ERROR:  array index must resolve to an integer value
psql-13.9-5432-pgsql=# SELECT * FROM cypher('graph',$$ MATCH ()<-[x]-() RETURN x.name $$) as (a agtype);
                   a
----------------------------------------


 {"last": "Gemignani", "first": "John"}
(3 rows)

psql-13.9-5432-pgsql=#
```

**Comment by MuhammadTahaNaveed** [comment]
But still, Shouldn't this work? 
```
SELECT * FROM cypher('graph',$$MATCH ()<-[x *]-() RETURN x[0].n0$$) as (a agtype);
```

P.S works in neo4j

**Comment by jrgemignani** [comment]
The edges are each wrapped in an array. All scalars are wrapped in a 1 element array and flagged as a scalar. Why that part for edges isn't working, I'm not sure yet. Btw, I have just created a PR #1036 to help with debugging types.

**Comment by MuhammadTahaNaveed** [comment]
oh... I thought it was an issue with agtype_access_operator. Have a look at PR #1034.

 > All scalars are wrapped in a 1 element array and flagged as a scalar.
 
And also, for the case of vle, what if length of path is greater than 1? Shouldn't two objects(edges) get wrapped into one array (as it is happening now)? for example
```
age=# SELECT * FROM cypher('new',$$MATCH ()<-[x *2]-() RETURN x$$) as (a agtype);
                                                                                                                           a                                                                  
                                                          
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------
 [{"id": 844424930131969, "label": "KNOWS", "end_id": 281474976710659, "start_id": 281474976710658, "properties": {}}::edge, {"id": 844424930131970, "label": "KNOWS", "end_id": 2814749767106
58, "start_id": 281474976710657, "properties": {}}::edge]
```

**Comment by jrgemignani** [comment]
I'm looking at it in the debugger. Apparently, the regular match and vle match go through different routes - each creates the edge in a different way.

**Comment by jrgemignani** [comment]
The regular match uses `_agtype_build_edge`. The vle match uses `agtype_value_build_edge`. Both convert their contents to agtype, but differently.

**Comment by jrgemignani** [comment]
@MuhammadTahaNaveed Actually, you are correct. The edges are returned in a list of edges for vle matches. It is just that in this case, there is only one edge in the path, in that list.

Here is a list of 2 -

```
psql-13.9-5432-pgsql=# SELECT * FROM cypher('graph',$$ MATCH ()-[e*2..2]->() RETURN e $$) as (a agtype);
                                                                                                                                   a

------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------
 [{"id": 1125899906842626, "label": "knows", "end_id": 281474976710664, "start_id": 281474976710663, "properties": {"id": 0}}::edge,
 {"id": 1125899906842625, "label": "knows", "end_id": 281474976710665, "start_id": 281474976710664, "properties": {"id": 1}}::edge]
(1 row)

psql-13.9-5432-pgsql=#
```

**Comment by jrgemignani** [comment]
It looks like this is a special case and maybe the access operator needs to be adjusted. Or, what is stored needs to be. I'm just not sure which, yet.

**Comment by jrgemignani** [comment]
I have a fix that I'm testing now.

**Comment by MuhammadTahaNaveed** [comment]
> It looks like this is a special case and maybe the access operator needs to be adjusted. Or, what is stored needs to be. I'm just not sure which, yet.

@jrgemignani oh okay! Actually I tried a fix for it in pr #1034. Can you give a look ? Thanks.

**Comment by jrgemignani** [comment]
@MuhammadTahaNaveed Yeah, we kind of both came up with a solution and PR at the same time. We need to coordinate better, lol.
